linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala

```diff
@@ -0,0 +1,458 @@
+        allFieldDefs,
+        superClass,
+        classImplementsAnyInterface,
+        clazz.hasInstances,
+        !clazz.hasDirectInstances,
+        hasRuntimeTypeInfo,
+        clazz.jsNativeLoadSpec,
+        clazz.jsNativeMembers.map(m => m.name.name -> m.jsNativeLoadSpec).toMap,
+        staticFieldMirrors
+      )
+    }
+
+    // Update specialInstanceTypes for ancestors of hijacked classes
+    if (clazz.kind == ClassKind.HijackedClass) {
+      def addSpecialInstanceTypeOnAllAncestors(jsValueType: Int): Unit =
+        strictClassAncestors.foreach(previousClassInfos(_).addSpecialInstanceType(jsValueType))
+
+      clazz.className match {
+        case BoxedBooleanClass =>
+          addSpecialInstanceTypeOnAllAncestors(JSValueTypeFalse)
+          addSpecialInstanceTypeOnAllAncestors(JSValueTypeTrue)
+        case BoxedStringClass =>
+          addSpecialInstanceTypeOnAllAncestors(JSValueTypeString)
+        case BoxedDoubleClass =>
+          addSpecialInstanceTypeOnAllAncestors(JSValueTypeNumber)
+        case BoxedUnitClass =>
+          addSpecialInstanceTypeOnAllAncestors(JSValueTypeUndefined)
+        case _ =>
+          ()
+      }
+    }
+
+    classInfo
+  }
+
+  /** Collects virtual and interface method calls.
+   *
+   *  That information will be used to decide what entries are necessary in
+   *  vtables and itables.
+   *
+   *  TODO Arguably this is a job for the `Analyzer`.
+   */
+  private class AbstractMethodCallCollector(classInfos: Map[ClassName, ClassInfo])
+      extends Traversers.Traverser {
+    def collectAbstractMethodCalls(clazz: LinkedClass): Unit = {
+      for (method <- clazz.methods)
+        traverseMethodDef(method)
+      for (jsConstructor <- clazz.jsConstructorDef)
+        traverseJSConstructorDef(jsConstructor)
+      for (export <- clazz.exportedMembers)
+        traverseJSMethodPropDef(export)
+    }
+
+    def collectAbstractMethodCalls(tle: LinkedTopLevelExport): Unit = {
+      tle.tree match {
+        case TopLevelMethodExportDef(_, jsMethodDef) =>
+          traverseJSMethodPropDef(jsMethodDef)
+        case _ =>
+          ()
+      }
+    }
+
+    override def traverse(tree: Tree): Unit = {
+      super.traverse(tree)
+
+      tree match {
+        case Apply(flags, receiver, methodName, _) if !methodName.name.isReflectiveProxy =>
+          receiver.tpe match {
+            case ClassType(className) =>
+              val classInfo = classInfos(className)
+              if (classInfo.hasInstances)
+                classInfo.registerDynamicCall(methodName.name)
+            case AnyType =>
+              classInfos(ObjectClass).registerDynamicCall(methodName.name)
+            case _ =>
+              // For all other cases, including arrays, we will always perform a static dispatch
+              ()
+          }
+
+        case _ =>
+          ()
+      }
+    }
+  }
+
+  /** Group interface types and types that implement any interfaces into buckets,
```

Comment:
@tanishiking thank you for writing this explanation. Otherwise this would have been much harder to understand.